### ISSUE TITLE
subquery issue for Delete Vertex Command

### ISSUE THREAD

**Issue by martingg88** [orientechnologies/orientdb#4523]
the following sql doesn't work for  2.1-rc5

DELETE VERTEX V FROM (SELECT EXPAND(OUT("target")) FROM #45:2)

**Comment by lvca** [comment]
Tested and works (Whisky database). Do you have any errors?

**Comment by martingg88** [comment]
no.. i do have this error. 

com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Encountered " <FROM> "FROM "" at line 1, column 17.
 Was expecting one of:
 <EOF> 
 <WHERE> ...
 <LIMIT> ...
 <RETURN> ...
 ";" ...

**Comment by lvca** [comment]
@luigidellaquila can you reproduce that? I tried with last 2.1 snapshot and works.

**Comment by martingg88** [comment]
this query work before 2.1-rc3, but no for 2.1-rc5 version

**Comment by martingg88** [comment]
is this the bug?

**Comment by nagarajasr** [comment]
@martingg88 i tested in 2.1-rc1, rc3 and rc6 and my observation is that this statement is not working in any of these releases. it seems to me a bug and i will investigate a solution, but will wait for @lvca to confirm that it is indeed a bug.

**Comment by martingg88** [comment]
okok. thanks

**Comment by martingg88** [comment]
any update on this?

**Comment by martingg88** [comment]
i guess the bug will occur if you try to delete from subquery that return empty record. Please help to solve this bug as soon as possible. thanks.

**Comment by johansjob** [comment]
Any updates? I have this problem as well but only for the new query parser. I'm using orientdb 2.1.0

I also tried an even simpler query:
DELETE VERTEX FROM (SELECT FROM #214:2392)

But I keep getting the same exception:

com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Encountered " <FROM> "FROM "" at line 1, column 15.
 Was expecting one of:
 \<TO> ...
 \<VALUES> ...
 \<SET> ...
 \<ADD> ...
 \<PUT> ...
 \<MERGE> ...
 \<CONTENT> ...
 \<REMOVE> ...
 \<ORDER> ...
 \<GROUP> ...
 \<OFFSET> ...
 \<RECORD> ...
 \<LUCENE> ...
 \<NEAR> ...
 \<WITHIN> ...
 \<THIS> ...
 \<RECORD_ATTRIBUTE> ...
 \<INTEGER_LITERAL> ...
 "(" ...
 "[" ...
 "`" ...
 "?" ...
 ":" ...
 "-" ...
 "*" ...
 \<IN> ...
 \<KEY> ...
 \<CLUSTER> ...
 \<IDENTIFIER> ...
 \<INDEX_IDENTIFIER> ...
 \<INDEXVALUES_IDENTIFIER> ...
 \<INDEXVALUESASC_IDENTIFIER> ...
 \<INDEXVALUESDESC_IDENTIFIER> ...
 \<CLUSTER_IDENTIFIER> ...
 \<METADATA_IDENTIFIER> ...
 "#" ...

**Comment by dcarr178** [comment]
I confirm this is happening to me as well in 2.1.1. I can't get delete vertex to work whether my subquery has results or not. 

```

orientdb {db=zumint_prod}> select from tEvent order by @rid desc limit 1

----+----------+------+-----------------+------+-------------------+-------------------+-----------------+-------------+----------------
#   |@RID      |@CLASS|Type             |Status|DateCreated        |DateUpdated        |out_rCompanyEvent|in_rCreatedBy|out_rEventObject
----+----------+------+-----------------+------+-------------------+-------------------+-----------------+-------------+----------------
0   |#39:159237|tEvent|NewProductListing|New   |2015-09-04 09:00:50|2015-09-04 09:00:50|[size=1]         |[size=1]     |[size=1]
----+----------+------+-----------------+------+-------------------+-------------------+-----------------+-------------+----------------

1 item(s) found. Query executed in 0.04 sec(s).
orientdb {db=zumint_prod}> delete vertex from (select from tEvent order by @rid desc limit 1)

Error: com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Encountered " <FROM> "from "" at line 1, column 15.
Was expecting one of:
    <TO> ...
    <VALUES> ...
    <SET> ...
    <ADD> ...
    <PUT> ...
    <MERGE> ...
    <CONTENT> ...
    <REMOVE> ...
    <ORDER> ...
    <GROUP> ...
    <OFFSET> ...
    <RECORD> ...
    <LUCENE> ...
    <NEAR> ...
    <WITHIN> ...
    <EXCEPTION> ...
    <THIS> ...
    <RECORD_ATTRIBUTE> ...
    <INTEGER_LITERAL> ...
    "(" ...
    "[" ...
    "?" ...
    ":" ...
    "-" ...
    "*" ...
    <IN> ...
    <KEY> ...
    <CLUSTER> ...
    <IDENTIFIER> ...
    <QUOTED_IDENTIFIER> ...
    <INDEX_IDENTIFIER> ...
    <INDEXVALUES_IDENTIFIER> ...
    <INDEXVALUESASC_IDENTIFIER> ...
    <INDEXVALUESDESC_IDENTIFIER> ...
    <CLUSTER_IDENTIFIER> ...
    <METADATA_IDENTIFIER> ...
    "#" ...


orientdb {db=zumint_prod}> delete vertex (select from tEvent order by @rid desc limit 1)

Error: com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Class 'SELECT' was not found

orientdb {db=zumint_prod}>
```

The command delete vertex from (<subquery>) used t work fine in version 2.0.12.

**Comment by luigidellaquila** [comment]
Hi, 

this is definitely a bug, I'm working on it right now.

**Comment by dcarr178** [comment]
Thanks!

**Comment by luigidellaquila** [comment]
fixed in branch 2.1.x.

The syntax 
```sql
DELETE VERTEX FROM (SELECT FROM FOO)
```
is now correct.

The following is incorrect anyway
```sql
DELETE VERTEX V FROM (SELECT FROM FOO)
```
because the statement can have only one target, and in this case V and the subquery would be both targets.

Keeping open because it has to be ported to 2.2 (develop branch)

**Comment by dcarr178** [comment]
@luigidellaquila Can you check deleting edges with subqueries as well? I'm not getting a query parse error but delete edge isn't working for me.

```
orientdb {db=zumint_prod}> select expand(inE('rBrandChannelProduct')) from tChannelProduct where BrandName <> in('rBrandChannelProduct')[0].Name limit 10

----+-----------+--------------------+---------+------+-------------------+-------------------
#   |@RID       |@CLASS              |out      |in    |DateCreated        |DateUpdated
----+-----------+--------------------+---------+------+-------------------+-------------------
0   |#23:2211542|rBrandChannelProduct|#14:38671|#19:6 |2015-09-04 14:41:13|2015-09-04 14:41:13
1   |#23:2259239|rBrandChannelProduct|#14:38671|#19:7 |2015-09-04 15:04:48|2015-09-04 15:04:48
2   |#23:2176881|rBrandChannelProduct|#14:23995|#19:8 |2015-09-04 14:24:08|2015-09-04 14:24:08
3   |#23:2194556|rBrandChannelProduct|#14:23995|#19:9 |2015-09-04 14:32:49|2015-09-04 14:32:49
4   |#23:2220776|rBrandChannelProduct|#14:23995|#19:10|2015-09-04 14:45:55|2015-09-04 14:45:55
5   |#23:2149044|rBrandChannelProduct|#14:23995|#19:11|2015-09-04 14:10:27|2015-09-04 14:10:27
6   |#23:2216202|rBrandChannelProduct|#14:23995|#19:12|2015-09-04 14:43:32|2015-09-04 14:43:32
7   |#23:2162396|rBrandChannelProduct|#14:14333|#19:13|2015-09-04 14:17:07|2015-09-04 14:17:07
8   |#23:2140265|rBrandChannelProduct|#14:14333|#19:14|2015-09-04 14:06:13|2015-09-04 14:06:13
9   |#23:2185161|rBrandChannelProduct|#14:14333|#19:16|2015-09-04 14:28:11|2015-09-04 14:28:11
----+-----------+--------------------+---------+------+-------------------+-------------------

10 item(s) found. Query executed in 0.041 sec(s).
orientdb {db=zumint_prod}> delete edge from (select expand(inE('rBrandChannelProduct')) from tChannelProduct where BrandName <> in('rBrandChannelProduct')[0].Name limit 10)

Delete record(s) '0' in 0.041000 sec(s).

orientdb {db=zumint_prod}>
```

Found records = 10 (per limit). Deleted records = 0.

**Comment by dcarr178** [comment]
Delete edge with limit doesn't work:

```
orientdb {db=zumint_prod}> select from rBrandChannelMerchant limit 10                                                                         
                                                                                                                                              
----+-----+---------------------+---------+-------+-------------------+-------------------                                                    
#   |@RID |@CLASS               |out      |in     |DateCreated        |DateUpdated                                                            
----+-----+---------------------+---------+-------+-------------------+-------------------                                                    
0   |#32:0|rBrandChannelMerchant|#21:15028|#14:0  |2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
1   |#32:1|rBrandChannelMerchant|#21:16   |#14:0  |2015-09-03 13:07:55|2015-09-03 13:07:55                                                    
2   |#32:2|rBrandChannelMerchant|#21:1910 |#14:0  |2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
3   |#32:3|rBrandChannelMerchant|#21:8271 |#14:0  |2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
4   |#32:4|rBrandChannelMerchant|#21:15   |#14:0  |2015-09-03 13:07:55|2015-09-03 13:07:55                                                    
5   |#32:5|rBrandChannelMerchant|#21:8    |#14:0  |2015-09-03 13:07:55|2015-09-03 13:07:55                                                    
6   |#32:6|rBrandChannelMerchant|#21:8914 |#14:352|2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
7   |#32:7|rBrandChannelMerchant|#21:9005 |#14:352|2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
8   |#32:8|rBrandChannelMerchant|#21:9044 |#14:352|2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
9   |#32:9|rBrandChannelMerchant|#21:8944 |#14:352|2015-09-03 13:07:55|2015-09-03 14:20:35                                                    
----+-----+---------------------+---------+-------+-------------------+-------------------                                                    
                                                                                                                                              
10 item(s) found. Query executed in 0.038 sec(s).                                                                                             
orientdb {db=zumint_prod}> delete edge rBrandChannelMerchant limit 10                                                                         
                                                                                                                                              
Error: com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Class 'LIMIT' was not found
                                                                                                                                              
orientdb {db=zumint_prod}>                                                                                                                    
```

**Comment by luigidellaquila** [comment]
Hi @dcarr178 

please check the documentation http://orientdb.com/docs/last/SQL-Delete-Edge.html
FROM word checks the starting point of the edge, not the edge itself (see FROM - TO in the docs), so this query is not correct:

```sql
delete edge from (select expand(inE('rBrandChannelProduct')) from tChannelProduct where BrandName <> in('rBrandChannelProduct')[0].Name limit 10)
```

you have to rewrite it this way:

```sql
delete edge E where @rid in (select expand(inE('rBrandChannelProduct')) from tChannelProduct where BrandName <> in('rBrandChannelProduct')[0].Name limit 10)
```

About the problem with LIMIT, I'm checking it right now

Thanks

Luigi

**Comment by luigidellaquila** [comment]
fixed LIMIT problem in branch 2.1.x, porting to 2.2 (develop branch)

**Comment by luigidellaquila** [comment]
fixed in both 2.1.x and develop branches. Closing
